src/clients/rust/build.rs

```diff
@@ -0,0 +1,158 @@
+        println!("cargo:rerun-if-changed={libdir}/lib{libname}.a");
+    } else if windows {
+        println!("cargo:rerun-if-changed={libdir}/{libname}.lib");
+    } else {
+        todo!();
+    }
+
+    let tb_client_h_path = format!("{cargo_manifest_dir}/assets/tb_client.h");
+    let bindings = bindgen::Builder::default()
+        .header(tb_client_h_path)
+        // Tell cargo to invalidate the built crate whenever any of the
+        // included header files changed.
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
+        .generate()
+        .context("Unable to generate bindings")?;
+
+    // Write the bindings to the $OUT_DIR/bindings.rs file.
+    let bindings_path = format!("{cargo_manifest_dir}/src/tb_client.rs");
+    bindings
+        .write_to_file(bindings_path)
+        .context("Couldn't write bindings!")?;
+
+    Ok(())
+}
+
+fn prepare_dependencies(manifest_dir: &str) -> AnyResult<()> {
+    let build_in_tree = is_build_in_tree(manifest_dir)?;
+    if build_in_tree {
+        build_tigerbeetle(manifest_dir)?;
+
+        let tb_client_h = format!("{manifest_dir}/../c/tb_client.h");
+        let tb_client_libs_dir = format!("{manifest_dir}/../c/lib");
+        let tb_client_dest = format!("{manifest_dir}/assets");
+        let tb_client_libs_dest = format!("{manifest_dir}/assets/lib");
+        let tb_client_h_dest = format!("{manifest_dir}/assets/tb_client.h");
+
+        fs::create_dir_all(&tb_client_dest)?;
+        fs::copy(&tb_client_h, tb_client_h_dest)?;
+
+        copy_dir_recursive(
+            Path::new(&tb_client_libs_dir),
+            Path::new(&tb_client_libs_dest),
+        )?;
+        emit_tigerbeetle_rerun_if_changed(manifest_dir)?;
+
+        Ok(())
+    } else {
+        todo!()
+    }
+}
+
+fn is_build_in_tree(manifest_dir: &str) -> AnyResult<bool> {
+    Ok(fs::exists(format!("{manifest_dir}/../../../build.zig"))?)
+}
+
+fn build_tigerbeetle(manifest_dir: &str) -> AnyResult<()> {
+    assert!(is_build_in_tree(manifest_dir)?);
+
+    let tigerbeetle_root = format!("{manifest_dir}/../../..");
+    let zig_compiler = if cfg!(unix) {
+        format!("{tigerbeetle_root}/zig/zig")
+    } else if cfg!(windows) {
+        format!("{tigerbeetle_root}/zig/zig.exe")
+    } else {
+        todo!()
+    };
+
+    if !fs::exists(&zig_compiler)? {
+        println!("cargo:warning=No zig compiler found at {zig_compiler}.");
+        println!("cargo:warning=You may need to run zig/download.sh (or zig/download.bat).");
+        panic!("No zig compiler found.");
+    }
+
+    let build_targets = [
+        "clients:c", // Build the tb_client library
+        "install",   // Build tigerbeetle binary for testing
+    ];
+
+    for build_target in build_targets {
+        let mut cmd = std::process::Command::new(&zig_compiler);
+        cmd.args(["build", build_target]);
+        let res = cmd.status()?;
+
+        if !res.success() {
+            panic!("zig build failed");
+        }
+    }
```

Comment:
For deployment to crates.io, we should package pre-build .so. That's not how Rust usually works, but, given that Zig itself is unstable, we should try to avoid running the Zig compiler on user's machine.

Reply:
Agreed about packaging the libraries for crates.io. There are some `todo!` codepaths for this.

I planned to package static libs for this, but you are suggesting dynamic. Why would we want to use dynamic libs here? It will force users to copy them around.

Reply:
> I planned to package static libs for this, but you are suggesting dynamic. 

I think I am just wrong here, using static is the way to go!